File: iosApp/iosApp/Modules/Authenticate/AuthenticateInteractor.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

protocol AuthenticateInteractor: AnyObject {
    func canUnlockWithBio(_ keyStoreItem: KeyStoreItem) -> Bool
    func unlockWithBiometrics(
        _ item: KeyStoreItem,
        title: String,
        handler: @escaping (Result<(String, String), Error>) -> Void
    )
    func isValid(item: KeyStoreItem, password: String, salt: String) -> Bool
}

// MARK: - DefaultAuthenticateInteractor

class DefaultAuthenticateInteractor {

    private var keyStoreService: KeyStoreService

    init(keyStoreService: KeyStoreService) {
        self.keyStoreService = keyStoreService
    }
}

// MARK: - DefaultAuthenticateInteractor

extension DefaultAuthenticateInteractor: AuthenticateInteractor {

    func canUnlockWithBio(_ keyStoreItem: KeyStoreItem) -> Bool {
        keyStoreItem.canUnlockWithBio() && keyStoreService.biometricsSupported()
    }

    func unlockWithBiometrics(
        _ item: KeyStoreItem,
        title: String,
        handler: @escaping (Result<(String, String), Error>) -> Void
    ) {
        keyStoreService.biometricsAuthenticate(
            title: title,
            handler: { success, error in
                DispatchQueue.main.async {
                    if success.boolValue {
                        let pass = self.keyStoreService.password(item: item) ?? ""
                        handler(.success((pass, "")))
                    } else {
                        handler(.failure(error ?? GlobalError.unknown))
                    }
                }
            }
        )
    }

    func isValid(item: KeyStoreItem, password: String, salt: String) -> Bool {
        guard let secretStorage = keyStoreService.secretStorage(
            item: item,
            password: password
        ) else {
            return false
        }

        return (try? secretStorage.decrypt(password: password)) != nil
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/authenticate/AuthenticateInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.authenticate

import com.sonsofcrypto.web3lib.services.keyStore.SignerStoreItem
import com.sonsofcrypto.web3lib.services.keyStore.SignerStoreService

interface AuthenticateInteractor {
    fun keyStoreItem(): SignerStoreItem?
    fun canUnlockWithBio(signerStoreItem: SignerStoreItem): Boolean
    fun unlockWithBiometrics(
        item: SignerStoreItem,
        title: String,
        handler: (AuthenticateData?, Error?) -> Unit
    )
    fun isValid(item: SignerStoreItem, password: String, salt: String): Boolean
}

class DefaultAuthenticateInteractor(
    private val signerStoreService: SignerStoreService
): AuthenticateInteractor {

    override fun keyStoreItem(): SignerStoreItem? = signerStoreService.selected

    override fun canUnlockWithBio(signerStoreItem: SignerStoreItem): Boolean =
        signerStoreItem.canUnlockWithBio() && signerStoreService.biometricsSupported()

    override fun unlockWithBiometrics(
        item: SignerStoreItem,
        title: String,
        handler: (AuthenticateData?, Error?) -> Unit
    ) {
        signerStoreService.biometricAuthenticate(title) { success: Boolean, error: Error? ->
            if (success) {
                val password = signerStoreService.password(item) ?: ""
                handler(AuthenticateData(password, ""), null)
            } else {
                handler(null, error ?: Error("Unknown error"))
            }
        }
    }

    override fun isValid(item: SignerStoreItem, password: String, salt: String): Boolean {
        return try {
            val secretStorage = signerStoreService.secretStorage(item, password) ?: return false
            secretStorage.decrypt(password)
            true
        } catch (error: Throwable) {
            false
        }
    }
}
